corehq/apps/reports/standard/tableau.py

```diff
@@ -63,34 +64,47 @@ def page_context(self):
-
-        if self.visualization.server.validate_hostname != '':
-            request = requests.Session()
-            request.mount('https://', host_header_ssl.HostHeaderSSLAdapter())
-            tabserver_response = request.post(tabserver_url,
-                                              post_arguments,
-                                              headers={'Host': self.visualization.server.validate_hostname})
-        else:
-            tabserver_response = requests.post(tabserver_url, post_arguments)
-
-        if tabserver_response.status_code == 200:
-            if tabserver_response.content != b'-1':
-                context.update({'ticket': tabserver_response.content.decode('utf-8')})
-                return render(self.request, self.template_name, context)
-            else:
-                context.update({"failure_message": _("Tableau trusted authentication failed")})
-                return render(self.request, 'reports/tableau_server_request_failed.html', context)
+
+@require_POST
+def tableau_visualisation_ajax(request, domain):
+    from requests_toolbelt.adapters import host_header_ssl
+    visualiation_data = {data: value[0] for data, value in dict(request.POST).items()}
+    server_name = visualiation_data.pop('server_name')
+    validate_hostname = visualiation_data.pop('validate_hostname')
+    target_site = visualiation_data.pop('target_site')
+    domain_username = visualiation_data.pop('domain_username')
+
+    tabserver_url = 'https://{}/trusted/'.format(server_name)
+    post_arguments = {'username': domain_username}
+
+    if target_site != 'Default':
+        post_arguments.update({'target_site': target_site})
+
+    if validate_hostname != '':
+        request = requests.Session()
+        request.mount('https://', host_header_ssl.HostHeaderSSLAdapter())
+        tabserver_response = request.post(tabserver_url,
+                                          post_arguments,
+                                          headers={'Host': validate_hostname})
+    else:
+        tabserver_response = requests.post(tabserver_url, post_arguments)
+
+    if tabserver_response.status_code == 200:
+        if tabserver_response.content != b'-1':
+            return JsonResponse({
+                "success": True,
+                "ticket": tabserver_response.content.decode('utf-8'),
+            })
         else:
-            message = _("Request to Tableau failed with status code {}").format(tabserver_response.status_code)
-            context.update({"failure_message": message})
-            return render(self.request, 'reports/tableau_server_request_failed.html', context)
+            return JsonResponse({
+                "success": False,
+                "message": _("Tableau trusted authentication failed"),
+            })
+    else:
+        return JsonResponse({
+            "success": False,
+            "message": _("Request to Tableau failed with status code {}").format(tabserver_response.status_code),
```

Comment:
Maybe out of scope here, but do you know if these failed requests will include more information in the response body?  I wonder if it's worth logging them somewhere or showing them to the user.

Reply:
I think is is valid and would be helpful but may require a small followup PR